core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/checkout/service/workflow/ValidateAvailabilityActivity.java

```diff
@@ -0,0 +1,86 @@
+ * (the "Commercial License" located at http://license.broadleafcommerce.org/commercial_license-1.1.txt)
+ * shall apply.
+ * 
+ * Alternatively, the Commercial License may be replaced with a mutually agreed upon license (the "Custom License")
+ * between you and Broadleaf Commerce. You may not use this file except in compliance with the applicable license.
+ * #L%
+ */
+package org.broadleafcommerce.core.checkout.service.workflow;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.broadleafcommerce.core.catalog.domain.Sku;
+import org.broadleafcommerce.core.inventory.service.ContextualInventoryService;
+import org.broadleafcommerce.core.order.domain.BundleOrderItem;
+import org.broadleafcommerce.core.order.domain.DiscreteOrderItem;
+import org.broadleafcommerce.core.order.domain.Order;
+import org.broadleafcommerce.core.order.domain.OrderItem;
+import org.broadleafcommerce.core.workflow.BaseActivity;
+import org.broadleafcommerce.core.workflow.ProcessContext;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.Resource;
+
+/**
+ * This will check the availability and quantities (if applicable) all order items in checkout request.
+ * Very similar to the {@link CheckUpdateAvailabilityActivity} but in the blCheckoutWorkflow instead.
+ * This should prevent succeed checkout, in case, when Sku became unavailable in range
+ * after it was adding to the cart and before completing the order.
+ */
+@Component("blValidateAvailabilityActivity")
+public class ValidateAvailabilityActivity extends BaseActivity<ProcessContext<CheckoutSeed>> {
+
+    public static final int ORDER = 750;
+    private static final Log LOG = LogFactory.getLog(ValidateAvailabilityActivity.class);
+
+    @Resource(name = "blInventoryService")
+    protected ContextualInventoryService inventoryService;
+
+    public ValidateAvailabilityActivity() {
+        setOrder(ORDER);
+    }
+
+    @Override
+    public ProcessContext<CheckoutSeed> execute(ProcessContext<CheckoutSeed> context) throws Exception {
+        Order order = context.getSeedData().getOrder();
+        if (order == null) {
+            return context;
+        }
+
+        for (OrderItem orderItem : order.getOrderItems()) {
+            Sku sku;
+            if (orderItem instanceof DiscreteOrderItem) {
+                sku = ((DiscreteOrderItem) orderItem).getSku();
+            } else if (orderItem instanceof BundleOrderItem) {
+                sku = ((BundleOrderItem) orderItem).getSku();
+            } else {
+                LOG.warn("Could not check availability; did not recognize passed-in item " + orderItem.getClass().getName());
+                return context;
+            }
+
+            Integer requestedQuantity = orderItem.getQuantity();
+            inventoryService.checkSkuAvailability(order, sku, requestedQuantity);
+
+            Integer previousQty = orderItem.getQuantity();
```

Comment:
I believe this whole block following `inventoryService.checkSkuAvailability(order, sku, requestedQuantity);` is unnecessary, since `order.getOrderItems()` will include the child order items as well.

This block of code is logic duplicated from `CheckUpdateAvailabilityActivity`, the original block from `CheckUpdateAvailablilityActivity` is this:

```java
Order order = context.getSeedData().getOrder();
Integer requestedQuantity = request.getItemRequest().getQuantity();
checkSkuAvailability(order, sku, requestedQuantity);

Integer previousQty = orderItem.getQuantity();
for (OrderItem child : orderItem.getChildOrderItems()) {
    Sku childSku = ((DiscreteOrderItem) child).getSku();
    Integer childQuantity = child.getQuantity();
    childQuantity = childQuantity / previousQty;
    checkSkuAvailability(order, childSku, childQuantity * requestedQuantity);
}
```

In this usage for `CheckUpdateAvailabilityActivity`, `requestedQuantity` is a new, different value from `previousQty`, because the quantity is being _updated_. Therefore, the Sku availability for the newly scaled child quantity must be validated.

However, in `ValidateAvailabilityActivity`, the quantity is _not_ being updated, so `requestedQuantity` and `previousQty` are equal, so nothing is gained by the calculation of child quantity using both values. Therefore it shouldn't be a problem to remove this block.

Reply:
@epenning yes, you are right, this block is not needed. Also added additional active check.